libbpf-rs/src/skeleton.rs

```diff
@@ -313,3 +316,26 @@ impl<'a> Drop for ObjectSkeletonConfig<'a> {
         unsafe { Box::from_raw(self.inner.obj) };
     }
 }
+
+pub trait SkelBuilder<'a> {
+    type Output: OpenSkel;
+    fn open(self) -> Result<Self::Output>;
+    fn open_opts(self, open_opts: libbpf_sys::bpf_object_open_opts) -> Result<Self::Output>;
+    fn object_builder(&self) -> &ObjectBuilder;
+    fn object_builder_mut(&mut self) -> &mut ObjectBuilder;
+}
+
+pub trait OpenSkel {
+    type Output: Skel;
+    fn load(self) -> Result<Self::Output>;
+    fn open_object(&self) -> &OpenObject;
+    fn open_object_mut(&mut self) -> &mut OpenObject;
+}
+
+pub trait Skel {
+    fn attach(&mut self) -> Result<()> {
+        unimplemented!()
```

Comment:
Why do we need this default definition?

Reply:
When there is no eBPF program in bpf.c, libbpf-cargo will not generate the attach function, which will cause an error in compiling the test_skeleton_empty_source test, so the default implementation is provided here. The following code is a snippet of the attach function generated by libbpf-cargo
```rust
fn gen_skel_attach(skel: &mut String, object: &mut BpfObj, obj_name: &str) -> Result<()> {
    if ProgIter::new(object.as_mut_ptr()).next().is_none() {
        return Ok(());
    }

    write!(
    ...
```

Reply:
Interesting. I haven't looked into it in detail, but intuitively I'd say we should probably change that and generate the function unconditionally. But I guess we can do that as a follow-up, if you don't want to do it right now.